File: TanbalType/DetectorSelfTest.cs
namespace TanbalType;

/// <summary>Sanity checks for bidirectional layout detection at startup.</summary>
internal static class DetectorSelfTest
{
    private static readonly string[] EnglishOnFaLayout =
        // «id» عمداً در لیست نیست: روی حالت فارسی «هی» می‌شود که خودش کلمهٔ فارسی معتبری است
        ["hello", "book", "world", "class", "please", "computer", "good", "test", "name", "google",
         "login", "dashboard", "database", "application", "fararu.com", "app", "api",
         // نام سایت‌ها و کلمات محاوره‌ای که روی حالت فارسی تایپ شده‌اند
         "digikala", "youtube", "instagram", "aparat", "www.digikala.com", "salam", "mamnoon",
         // انگلیسی که قبلاً به‌خاطر تداخل با لغت‌نامه اصلاح نمی‌شد (واژهٔ فارسیِ مسدودکننده حذف شد)
         "just", "load", "can", "did",
         "as","at","co","he","had","how","if","ill","in","ir","is","it","its","idk","jan",
         "gg","gif","git","gn","go","nah","ok","on","or","os","pc","pdf","she","so","tbh",
         "thx","to","ty","ui","us","ux","why","wp"];

    private static readonly string[] PersianOnEnLayout =
        ["sghl", "ldo,hil", "vh", "fhc", ";vnd", "ffdkd", "ldai", "kldai", "dh", "fh", "hc",
         // کلمات محاوره‌ای فارسی که روی حالت انگلیسی تایپ شده‌اند
         "]x,vd" /* چطوری */, "]ofv" /* چخبر */, "pgi" /* حله */, "nd'i" /* دیگه */,
         "jaodwa" /* تشخیصش */, ";jhfa" /* کتابش */];

    private static readonly string[] PersianShouldNotFix =
        ["برنامه", "بعد", "خیلی", "به", "از", "نمایش", "اطلاعات", "مطالعه", "باید", "سلام", "که", "مدیر",
         "میشه", "نمیشه", "باشه", "دیگه", "حله", "مخم", "چاکرم", "اوکی", "قربونت", "دمتگرم",
         // کلمات لغت‌نامه‌ای با پسوند چسبان — نباید به انگلیسی تبدیل شوند
         "تشخیصش", "تشخیصشون", "برنامشون", "لپتاپم", "موبایلت", "پروژهام",
         // کلمات واقعی فارسی که با انگلیسی تداخل دارند ولی باید حفظ شوند
         "شد", "یخ", "هدف", "خب", "مخزن", "اثاث", "زهد", "مخل"];

    private static readonly string[] EnglishShouldNotFix =
        ["performance", "analysis", "project", "experience", "authentication", "shopping", "slam",
         "book", "id", "app", "api",
         // نام سایت‌های ایرانی که در انگلیسی معنایی ندارند ولی نباید اصلاح شوند
         "digikala", "aparat", "divar", "snapp", "torob", "sheypoor", "varzesh3", "filimo",
         "cafebazaar", "zarinpal", "www.aparat.com", "youtube", "instagram", "telegram",
         // کلمات محاوره‌ای چت و فینگلیش
         "lol", "btw", "thx", "okay", "yeah", "gonna", "salam", "mamnoon", "chetori", "kheili", "azizam"];

    public static void RunAndLog()
    {
        var failed = 0;

        foreach (var word in EnglishOnFaLayout)
        {
            var onScreen = Mapper.EnKeysToPersian(word);
            var got = Detector.DetectWrongLayout(onScreen, currentLayoutIsPersian: true);
            if (got != word)
            {
                AppLog.Write($"SelfTest FA->EN FAIL: {onScreen} -> '{got}' (expect {word})");
                failed++;
            }
        }

        foreach (var keys in PersianOnEnLayout)
        {
            var expected = Mapper.EnKeysToPersian(keys);
            var got = Detector.DetectWrongLayout(keys, currentLayoutIsPersian: false);
            if (got != expected)
            {
                AppLog.Write($"SelfTest EN->FA FAIL: {keys} -> '{got}' (expect {expected})");
                failed++;
            }
        }

        foreach (var word in PersianShouldNotFix)
        {
            if (Detector.DetectWrongLayout(word, currentLayoutIsPersian: true) is not null)
            {
                AppLog.Write($"SelfTest FA false-positive: {word}");
                failed++;
            }
        }

        foreach (var word in EnglishShouldNotFix)
        {
            if (Detector.DetectWrongLayout(word, currentLayoutIsPersian: false) is not null)
            {
                AppLog.Write($"SelfTest EN false-positive: {word}");
                failed++;
            }
        }

        if (failed == 0)
            AppLog.Write("SelfTest OK — corrections + no false positives");
        else
            AppLog.Write($"SelfTest FAILED — {failed} case(s)");
    }
}

File: TanbalType-macOS/Sources/DetectorSelfTest.swift
import Foundation

/// آزمون درستیِ تشخیص دوطرفه (پورت DetectorSelfTest.cs). داده‌ها از سورس C# تولید می‌شوند.
///
/// کلیدهای PersianOnEnLayout برای چیدمان ویندوز نوشته شده‌اند؛ برای چیدمان‌های مک، خودِ واژهٔ
/// فارسیِ مورد انتظار نگه داشته می‌شود و کلیدهایش از روی چیدمانِ آزمون‌شده دوباره ساخته می‌شود.
enum DetectorSelfTest {
    /// خروجی: تعداد موارد ناموفق.
    ///
    /// چیدمان‌های Persian-QWERTY (آوایی) و Persian (قدیمی) عمداً جزو آزمون پیش‌فرض نیستند: روی آن‌ها
    /// بسیاری از کلمات کوتاه انگلیسی خودشان واژهٔ فارسیِ معتبرند (مثلاً salam → «سالام»).
    @discardableResult
    static func run(layoutIDs: [String]) -> Int {
        let current = Mapper.sourceID
        var failed = 0

        Mapper.useWindowsLayout()
        let persianTargets = DetectorData.persianOnEnLayout.map(Mapper.enKeysToPersian)

        // نگاشت ویندوز همیشه آزموده می‌شود؛ به‌علاوهٔ چیدمان‌های مکِ خواسته‌شده
        var layouts: [(String, () -> Void)] = [("windows", { Mapper.useWindowsLayout() })]
        for id in layoutIDs where id != "windows" {
            guard let source = LayoutManager.source(withID: id) else { continue }
            layouts.append((id, { Mapper.configure(from: source) }))
        }

        for (name, activate) in layouts {
            activate()
            let before = failed
            failed += runCases(persianTargets: persianTargets)
            AppLog.write("SelfTest [\(name)]: \(failed - before) failure(s)")
        }

        if let source = LayoutManager.source(withID: current) {
            Mapper.configure(from: source)
        } else {
            Mapper.useWindowsLayout()
        }

        AppLog.write(failed == 0
            ? "SelfTest OK — corrections + no false positives"
            : "SelfTest FAILED — \(failed) case(s)")
        return failed
    }

    private static func runCases(persianTargets: [String]) -> Int {
        var failed = 0

        for word in DetectorData.englishOnFaLayout {
            let onScreen = Mapper.enKeysToPersian(word)
            let got = Detector.detectWrongLayout(onScreen, currentLayoutIsPersian: true)
            if got != word {
                AppLog.write("SelfTest FA->EN FAIL: \(onScreen) -> '\(got ?? "nil")' (expect \(word))")
                failed += 1
            }
        }

        for expected in persianTargets {
            let keys = Mapper.persianToEnKeys(expected)
            let got = Detector.detectWrongLayout(keys, currentLayoutIsPersian: false)
            if got != expected {
                AppLog.write("SelfTest EN->FA FAIL: \(keys) -> '\(got ?? "nil")' (expect \(expected))")
                failed += 1
            }
        }

        for word in DetectorData.persianShouldNotFix
        where Detector.detectWrongLayout(word, currentLayoutIsPersian: true) != nil {
            AppLog.write("SelfTest FA false-positive: \(word)")
            failed += 1
        }

        for word in DetectorData.englishShouldNotFix
        where Detector.detectWrongLayout(word, currentLayoutIsPersian: false) != nil {
            AppLog.write("SelfTest EN false-positive: \(word)")
            failed += 1
        }

        for word in DetectorData.siteNamesShouldNotFix {
            if let fixed = Detector.detectWrongLayout(word, currentLayoutIsPersian: false) {
                AppLog.write("SelfTest site-name interference: \(word) -> '\(fixed)'")
                failed += 1
            }
        }

        for word in DetectorData.digitTokensShouldNotFix
        where Detector.detectWrongLayout(word, currentLayoutIsPersian: true) != nil
            || Detector.detectWrongLayout(word, currentLayoutIsPersian: false) != nil {
            AppLog.write("SelfTest digit false-positive: \(word)")
            failed += 1
        }

        return failed
    }
}
